File: test/dummy/app/components/dummy/ui/boolean_toggle_component.rb
# frozen_string_literal: true

class Dummy::Ui::BooleanToggleComponent < Folio::Console::ApplicationComponent
  bem_class_name :label, :verbose

  def initialize(record: nil,
                 attribute:,
                 url: nil,
                 checked: nil,
                 disabled: false,
                 label: nil,
                 verbose: false,
                 as: nil,
                 confirm: false,
                 class_name: nil,
                 small_label: false)
    @record = record
    @attribute = attribute
    @disabled = disabled
    @label = label || small_label || verbose
    @verbose = verbose
    @as = as
    @confirm = confirm
    @url = url
    @class_name = class_name
    @small_label = small_label
    @checked = checked
  end

  def checked
    @checked.nil? ? !!@record.try(@attribute) : @checked
  end

  def name
    if @as || @record
      "#{as}[#{@attribute}]"
    else
      @attribute
    end
  end

  def as
    @as || @record.class.model_name.param_key
  end

  def verbose_label(boolean)
    @record.class.human_attribute_name("#{@attribute}/#{boolean}")
  end

  def label_string
    if @label == true
      @record.class.human_attribute_name(@attribute)
    else
      @label
    end
  end

  def confirmation
    if @confirm.is_a?(String)
      @confirm
    elsif @confirm
      "true"
    end
  end

  def data
    stimulus_controller("d-ui-boolean-toggle", values: {
      url: @url,
      confirmation:,
      attribute: @attribute,
    }, classes: %w[loading])
  end
end

File: test/dummy/app/components/dummy/ui/boolean_toggle_component.js
window.Folio.Stimulus.register('d-ui-boolean-toggle', class extends window.Stimulus.Controller {
  static values = {
    url: { type: String, default: '' },
    attribute: String,
    confirmation: { type: String, default: '' }
  }

  static classes = ['loading']

  onInput (e) {
    const input = e.target

    if (this.confirmationValue) {
      if (this.confirmationValue === 'true') {
        if (window.Folio.Confirm.confirm(() => { this.onInputInner(input) })) {
          return
        }
      } else {
        if (window.Folio.Confirm.message(() => { this.onInputInner(input) }, this.confirmationValue)) {
          return
        }
      }

      input.checked = !input.checked
      return
    }

    this.onInputInner(input)
  }

  onInputInner (input) {
    if (this.urlValue) {
      if (this.element.classList.contains(this.loadingClass)) return

      this.element.classList.add(this.loadingClass)

      const parts = input.name.replace(/\]/g, '').split('[')

      let data = { [parts[parts.length - 1]]: input.checked }

      for (let i = parts.length - 2; i >= 0; i -= 1) {
        data = { [parts[i]]: data }
      }

      data._trigger = 'd-ui-boolean-toggle'

      window.Folio.Api.apiPut(this.urlValue, data).then((res) => {
        this.element.classList.remove(this.loadingClass)

        if (res && res.data && res.data.f_c_catalogue_published_dates) {
          const row = this.element.closest('.f-c-catalogue__row')
          if (row) {
            const cell = row.querySelector('.f-c-catalogue__cell--published_dates .f-c-catalogue__cell-value')

            if (cell) {
              cell.innerHTML = res.data.f_c_catalogue_published_dates
            }
          }
        }
      }).catch((res) => {
        window.FolioConsole.Flash.alert(res.message)

        this.element.classList.remove(this.loadingClass)
        input.checked = !input.checked
      })
    }

    this.dispatch('changed', { detail: { attribute: this.attributeValue, checked: input.checked } })
  }
})
